homeassistant/components/smarttub/controller.py

```diff
@@ -0,0 +1,123 @@
+"""Interface to the SmartTub API."""
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from aiohttp import client_exceptions
+import async_timeout
+from smarttub import APIError, LoginFailed, SmartTub
+from smarttub.api import Account
+
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, POLLING_TIMEOUT, SCAN_INTERVAL
+from .helpers import get_spa_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmartTubController:
+    """Interface between Home Assistant and the SmartTub API."""
+
+    def __init__(self, hass):
+        """Initialize an interface to SmartTub."""
+        self._hass = hass
+        self._api = SmartTub(async_get_clientsession(hass))
+        self._account = None
+        self.spas = set()
+        self._spa_devices = {}
+
+        self.coordinator = None
+
+    async def async_setup_entry(self, entry):
+        """Perform initial setup.
+
+        Authenticate, query static state, set up polling, and otherwise make
+        ready for normal operations .
+        """
+
+        try:
+            self._account = await self.login(
+                entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD]
+            )
+        except (
```

Comment:
Might be cleaner to just trap the execption `        except LoginFailed:` here instead

Reply:
OK, I moved the exception handling out of `login` into this method and the config flow.